engine/execution/state/state.go

```diff
@@ -336,6 +336,22 @@ func (s *state) GetExecutionResultID(ctx context.Context, blockID flow.Identifie
 func (s *state) SaveExecutionResults(ctx context.Context, header *flow.Header, endState flow.StateCommitment,
 	chunkDataPacks []*flow.ChunkDataPack, executionReceipt *flow.ExecutionReceipt, events []flow.EventsList, serviceEvents flow.EventsList,
 	results []flow.TransactionResult) error {
+	err := s.saveExecutionResults(ctx, header, endState, chunkDataPacks, executionReceipt, events, serviceEvents, results, false)
+	if err == nil {
+		return nil
+	}
+
+	// if result already exists, we try again by force re-indexing the result by the block ID,
+	if storage.IsResultAlreadyExistsErr(err) {
```

Comment:
I find this error handling too coarse. I feel there are a lot of reasons why `ResultAlreadyExistsErr` could be returned, but I am not sure that forcing an overwrite in all cases is the appropriate response. Essentially, if you are responding to all cases with a forced override, you have to argue that this is the correct action in _all possible cases_ and that this behaviour is safe in a byzantine environment and won't get the EN accidentally slashed. 
To me the current approach feels like a "continue on a best effort basis", which is a pattern we are aiming to avoid. 

My assumption why you added this code is: 
* The EN maintains an index  BlockID -> resultID for its _own_ results.
* You don't clean this up in [`removeExecutionResultsFromHeight`](https://github.com/onflow/flow-go/blob/7b041a45902c4cce7fee7c293bab3711e6273fbb/cmd/util/cmd/rollback-executed-height/cmd/rollback-executed-height.go#L93), so when the EN later executes blocks again, it potentially finds that the block it just executed already has a resultID associated with it (specifically the result the EN had _before_ you removed the results). 




### Suggestion
* I would find it much cleaner if you also cleaned up this secondary index from BlockID -> resultID in [`removeExecutionResultsFromHeight`](https://github.com/onflow/flow-go/blob/7b041a45902c4cce7fee7c293bab3711e6273fbb/cmd/util/cmd/rollback-executed-height/cmd/rollback-executed-height.go#L93)
* thereby, this work-around due to an the incomplete cleanup would not be necessary.

Reply:
We need this check here, not because we didn't remove the index, it's because we can't remove the result due to the fact a result might be included in a block's payload already. Removing the result might cause EN unable to load this block for re-executing it. So the result can't be removed during rolling back.

The benefit of keeping the index is that, we can tell if we are executing a block for the first time or the second time by checking if this index exists. 

If we are executing for the first time, then we should not run into ResultAlreadyExist error. 
If we are executing for the second time, then there are still two cases: 
1. the produced result is the same result as the first time
2. the produced result is a different one

For the second case, we should not run into ResultAlreadyExist error, but for the first case, we will. 

Can we distinguish the two cases and only handle this error if it's the first case? 

Yes, but we have to query the result by ID before inserting the result. Unfortunately, we can't do it because saving the execution results is done with a batch operation. Badger doesn't support any DB read within a batch update( the [WriteBatch struct](https://github.com/dgraph-io/badger/blob/master/batch.go#L170) only has `Set` method, no `Get` method). 

One workaround is to query the index and result before batch inserting result related data. Given the safe assumption that index and result will not be removed during the node is running, the query result should still hold true when later batch operation is run in a separate DB transaction.

The only drawback is that the two additional query slows down the saving of the result, given this is very rare to happen. 

But even we add the query, I think it's essentially the same as with this check. My point is that adding the query could avoid this check, but they are handling (distinguishing) the same cases. So I'm not sure we would gain much. 

Thoughts?

Reply:
I think what you says makes sense @AlexHentschel 

We should remove all the results related data including events, service events, all kinds of index, as if the block was never executed. Except the result and the receipt itself, because it's possible they have been included in a block's payload.

I'm able to remove this check, since we actually won't hit this error, because the batch write would overwrite the result if already exists.